pkg/envoy/lds/builder.go

```diff
@@ -0,0 +1,599 @@
+
+	for i, f := range hb.filters {
+		if f.Name == envoy.HTTPRouterFilterName {
+			routerFilterIndex = i
+			break
+		}
+	}
+
+	// Guard against the case where multiple router filters are accidentally programmed.
+	// This panics to ensure this never happens in code.
+	if routerFilterIndex != -1 && filter.Name == envoy.HTTPRouterFilterName {
+		panic("multiple router filters not allowed")
+	}
+
+	if routerFilterIndex != lastIndex {
+		routerFilter := hb.filters[routerFilterIndex]
+		hb.filters = append(hb.filters[:routerFilterIndex], hb.filters[routerFilterIndex+1])
+		hb.filters = append(hb.filters, routerFilter)
+	}
+
+	return hb
+}
+
+// LocalReplyConfig sets the given LocalReplyConfig on the builder
+func (hb *httpConnManagerBuilder) LocalReplyConfig(config *xds_hcm.LocalReplyConfig) *httpConnManagerBuilder {
+	hb.localReplyConfig = config
+	return hb
+}
+
+// Tracing sets the Tracing config on the builder
+func (hb *httpConnManagerBuilder) Tracing(config *xds_hcm.HttpConnectionManager_Tracing) *httpConnManagerBuilder {
+	hb.tracing = config
+	return hb
+}
+
+// Build builds the HttpConnectionManager filter from the builder config
+func (hb *httpConnManagerBuilder) Build() (*xds_listener.Filter, error) {
+	connManager := &xds_hcm.HttpConnectionManager{
+		StatPrefix:  hb.statsPrefix,
+		CodecType:   xds_hcm.HttpConnectionManager_AUTO,
+		HttpFilters: hb.filters,
+		RouteSpecifier: &xds_hcm.HttpConnectionManager_Rds{
+			Rds: &xds_hcm.Rds{
+				ConfigSource:    envoy.GetADSConfigSource(),
+				RouteConfigName: hb.routeConfigName,
+			},
+		},
+		AccessLog: envoy.GetAccessLog(),
+		UpgradeConfigs: []*xds_hcm.HttpConnectionManager_UpgradeConfig{
+			{
+				UpgradeType: websocketUpgradeType,
+			},
+		},
+	}
+
+	if hb.tracing != nil {
+		connManager.GenerateRequestId = &wrappers.BoolValue{
+			Value: true,
+		}
+		connManager.Tracing = hb.tracing
+	}
+
+	marshalled, err := anypb.New(connManager)
+	if err != nil {
+		return nil, err
+	}
+
+	return &xds_listener.Filter{
+		Name:       envoy.HTTPConnectionManagerFilterName,
+		ConfigType: &xds_listener.Filter_TypedConfig{TypedConfig: marshalled},
+	}, nil
+}
+
+// TCPProxyBuilder returns a TCP proxy builder instance
+func TCPProxyBuilder() *tcpProxyBuilder { //nolint: revive // unexported-return
+	return &tcpProxyBuilder{}
+}
+
+// StatsPrefix sets the stats prefix to use for the TCP proxy filter
+func (tb *tcpProxyBuilder) StatsPrefix(statsPrefix string) *tcpProxyBuilder {
+	tb.statsPrefix = statsPrefix
+	return tb
+}
+
+// Cluster sets the cluster to use for the TCP proxy filter
+func (tb *tcpProxyBuilder) Cluster(cluster string) *tcpProxyBuilder {
+	tb.cluster = cluster
+	return tb
+}
+
+func (tb *tcpProxyBuilder) WeightedClusters(wc []service.WeightedCluster) *tcpProxyBuilder {
+	tb.weightedClusters = wc
+	return tb
+}
+
+// Build builds the TCP proxy filter
+func (tb *tcpProxyBuilder) Build() (*xds_listener.Filter, error) {
+	if tb.cluster != "" && len(tb.weightedClusters) > 0 {
```

Comment:
i would consider changing the builder to have only 1 way to accept clusters

`AddCluster(cluster string, weight int)`

Reply:
I prefer to keep this as is since I want to ensure you can set both a single cluster as well as weighted clusters through the builder.